### PR TITLE
Graphite: Expand metric names for variables

### LINKED ISSUE #2343: Templating: Regex on full graphite expended matches
I know this has been discussed in part in https://github.com/grafana/grafana/issues/2160:

I'm trying to limit a template variable query by including the value of a (independent) variable and then using _parts_ of the _full metric name_. Let's have an example:

Given metric names
city1.services.user1.bananas_eaten.count
city1.services.user1.apples_delivered.count
city1.gates.carts_passed.count

I want to capture template variables `$user` and `$city`, where grabbing possible user values is easy:
just a graphite query for `*.services.user*`. Now, I want `$city` to be limited by the `$user`. 

Here I'm stuck. I can ask the graphite API: `https://my.graphite/graphite/metrics/expand?query=city*.services.$user` and use a regex to pull out the values. But how do I do that in Grafana? All I can do is `find` metrics, which only gives me the last node of the metric name.

Can the template query mechanism somehow expose the full metric name so I could decide which values to construct out of it?

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

Variable query for Graphite allows to search for metrics, however it's quite limited because results contain possible values but **occurring only at the last level of the query** (please check the example below). This PR introduces "expand" function allowing to use expand API to get the full name of matching metrics. In combination with regex it will allow to extract any part of the metric name.

Example:

Metrics:
- `prod.servers.server001.cpu`
- `prod.servers.server002.cpu`
- `test.servers.server001.cpu`

Search results: (normal query is already available, expanded query support is added in this PR)

| normal query | results | expanded query | expanded results |
|--------------|---------|----------------|------------------|
| `*` | prod, test | `expand(*)` | prod, test
| `*.servers` | servers | `expand(*.servers)` | prod.servers, test.servers |
| `test.servers` | servers | `expand(test.servers)` | test.servers |
| `*.servers.*` | server001,server002 | `expand(*.servers.*)` | prod.servers.server001, prod.servers.server002, test.servers.server001 |
| `test.servers.*` | server001 | `expand(test.servers.*)` | test.servers.server001 |
| `*.servers.*.cpu` | cpu | `expand(*.servers.*.cpu)` | prod.servers.server001.cpu, prod.servers.server002.cpu, test.servers.server001.cpu |

⚠️ Note: "normal query" is the same as "expand query" with a regex matching the last part of the name.

Why is it important?

The available implementation allows to handle only simpler (though most common) cases where only the last part of the metric should become a variable value. However, it's not  possible extract other parts of metric names. 

Imagine you needed a dropdown to select the environment (`prod` or `test` in the example above). Normal search could capture the top level name but only of **all** metrics you have. Let's say besides `prod.*` and `test.*` metrics you had `statsd.*`. With normal query "statsd" would also be captured. With "expand" it's possible to write do this:

Query: `expand(*.servers)`
Regex: `/^(\w+).*/`

Expand will first find only metrics that have `servers` in the second level and then regex will capture only environment names. 

How to test it?

Run: `make devenv sources=graphite`
Send some sample data: https://gist.github.com/ifrost/ada07138775d5295fba169d27abddf8d  

Create a dashboard with following query: `$env.servers.$server.cpu`

Screenshot 2021-05-04 at 14 34 08

Create variables:

`env`:

Screenshot 2021-05-04 at 14 35 14

and `server`:

Screenshot 2021-05-04 at 14 35 24

**Which issue(s) this PR fixes**:

Fixes #2343 

**Special notes for your reviewer**: Same example is included in updated docs.

### CODE REVIEW

**Review by achatterjee-grafana — COMMENTED** [review]
Is this for V8.0?

**Comment by ifrost** [comment]
> Is this for V8.0?

Yes! Sorry, forgot about the milestone. Updated.

File: docs/sources/datasources/graphite.md [hunk 1]
```diff
@@ -141,6 +141,27 @@ For more details, see the [Graphite docs on the autocomplete API for tags](http:
...
+The results will contain all possible values occurring only at the last level of the query. To get full metric names matching the query 
+use expand function (`expand(*.servers.*)`). 
+
+#### Comparison between expanded and non-expanded metric search results
+
+Example metrics:
+- `prod.servers.001.cpu`
+- `prod.servers.002.cpu`
+- `test.servers.001.cpu`
+
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/graphite.md:153]:
```suggestion
The following examples show how expanded and non-expanded queries can be used to fetch specific parts of the metrics name.
```

File: docs/sources/datasources/graphite.md [hunk 2]
```diff
@@ -141,6 +141,27 @@ For more details, see the [Graphite docs on the autocomplete API for tags](http:
...
+| normal query | results | expanded query | expanded results |
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/graphite.md:154]:
```suggestion
| non-expanded query | results | expanded query | expanded results |
```

File: docs/sources/datasources/graphite.md [hunk 3]
```diff
@@ -141,6 +141,27 @@ For more details, see the [Graphite docs on the autocomplete API for tags](http:
...
+|--------------|---------|----------------|------------------|
+| `*` | prod, test | `expand(*)` | prod, test
+| `*.servers` | servers | `expand(*.servers)` | prod.servers, test.servers |
+| `test.servers` | servers | `expand(test.servers)` | test.servers |
+| `*.servers.*` | 001,002 | `expand(*.servers.*)` | prod.servers.001, prod.servers.002, test.servers.001 |
+| `test.servers.*` | 001 | `expand(test.servers.*)` | test.servers.001 |
+| `*.servers.*.cpu` | cpu | `expand(*.servers.*.cpu)` | prod.servers.001.cpu, prod.servers.002.cpu, test.servers.001.cpu |
+
+You can combine `expand` with a regex to match specific part of the metric name path.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/graphite.md:163]: Remove, as we covered this in the intro section I have suggested?

File: docs/sources/datasources/graphite.md [hunk 4]
```diff
@@ -141,6 +141,27 @@ For more details, see the [Graphite docs on the autocomplete API for tags](http:
...
+
+| normal query | results | expanded query | expanded results |
+|--------------|---------|----------------|------------------|
+| `*` | prod, test | `expand(*)` | prod, test
+| `*.servers` | servers | `expand(*.servers)` | prod.servers, test.servers |
+| `test.servers` | servers | `expand(test.servers)` | test.servers |
+| `*.servers.*` | 001,002 | `expand(*.servers.*)` | prod.servers.001, prod.servers.002, test.servers.001 |
+| `test.servers.*` | 001 | `expand(test.servers.*)` | test.servers.001 |
+| `*.servers.*.cpu` | cpu | `expand(*.servers.*.cpu)` | prod.servers.001.cpu, prod.servers.002.cpu, test.servers.001.cpu |
+
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/graphite.md:162]:
```suggestion
As you can see from the results, the non-expanded query is the same as an expanded query with a regex matching the last part of the name.
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added change suggestions.

File: docs/sources/datasources/graphite.md (hunk 3 above)
Comment (ifrost) [reply docs/sources/datasources/graphite.md]: Makes sense. Removed.

**Review by achatterjee-grafana — APPROVED** [review]
Doc looks ok, we can refine it further if needed.

**Review by gabor — APPROVED** [review]
looks good to me 👍 !

**Pull request merged**